modules/hardware/common/shared-mem.nix

```diff
@@ -0,0 +1,232 @@
+            config = {
+              microvm = {
+                qemu = {
+                  extraArgs = [
+                    "-device"
+                    "ivshmem-doorbell,vectors=${vectors},chardev=ivs_socket,flataddr=${config.ghaf.shm.flataddr}"
+                    "-chardev"
+                    "socket,path=${config.ghaf.shm.hostSocketPath},id=ivs_socket"
+                  ];
+                };
+                kernelParams = [ "kvm_ivshmem.flataddr=${config.ghaf.shm.flataddr}" ];
+              };
+              boot.extraModulePackages = [
+                (pkgs.linuxPackages.callPackage ../../../packages/memsocket/module.nix {
+                  inherit (config.microvm.vms.${vmName}.config.config.boot.kernelPackages) kernel;
+                  vmCount = config.ghaf.shm.instancesCount;
+                })
+              ];
+              services = {
+                udev = {
+                  extraRules = ''
+                    SUBSYSTEM=="misc",KERNEL=="ivshmem",GROUP="kvm",MODE="0666"
+                  '';
+                };
+              };
+              environment.systemPackages = optionals config.ghaf.shm.enable [
+                memsocket
+              ];
+              systemd.user.services.memsocket =
+                if vmName == "gui-vm" then
+                  lib.mkIf config.ghaf.shm.enable {
+                    enable = true;
+                    description = "memsocket";
+                    after = [ "labwc.service" ];
+                    serviceConfig = {
+                      Type = "simple";
+                      ExecStart = "${memsocket}/bin/memsocket -c ${config.ghaf.shm.clientSocketPath}";
+                      Restart = "always";
+                      RestartSec = "1";
+                    };
+                    wantedBy = [ "ghaf-session.target" ];
+                  }
+                else
+                  # machines connecting to gui-vm
+                  let
+                    vmIndex = lib.lists.findFirstIndex (vm: vm == vmName) null config.ghaf.shm.vms_enabled;
+                  in
+                  lib.mkIf config.ghaf.shm.enable {
+                    enable = true;
+                    description = "memsocket";
+                    serviceConfig = {
+                      Type = "simple";
+                      ExecStart = "${memsocket}/bin/memsocket -s ${config.ghaf.shm.serverSocketPath} ${builtins.toString vmIndex}";
+                      Restart = "always";
+                      RestartSec = "1";
+                    };
+                    wantedBy = [ "default.target" ];
+                  };
+            };
+          };
+        };
+      };
+    in
+    mkIf config.ghaf.shm.enable (
+      foldl' lib.attrsets.recursiveUpdate { } (map makeAssignment config.ghaf.shm.vms_enabled)
+    );
+
+  config.ghaf.hardware.definition.gpu.kernelConfig.kernelParams = optionals config.ghaf.shm.enable [
+    "kvm_ivshmem.flataddr=${config.ghaf.shm.flataddr}"
+  ];
```

Comment:
Could you use `config.microvm.vms.gui-vm.config.config.boot.kernelParams` directly instead of the hardware configuration?